bundle/edu.gemini.horizons.server/src/main/scala/edu/gemini/horizons/server/backend/HS2.scala

```diff
@@ -0,0 +1,299 @@
+              List(Row(HorizonsDesignation.AsteroidNewStyle(m.group(2)) : HorizonsDesignation.Asteroid, m.group(1)))
+            } \/> "Could not match '90377 Sedna (2003 VB12)' header pattern."
+
+          // Single result with form: JPL/HORIZONS      4 Vesta     2015-Dec-31 11:40:21
+          lazy val case2 =
+            """  +(\d+) ([^(]+?)  """.r.findFirstMatchIn(header).map { m =>
+              List(Row(HorizonsDesignation.AsteroidOldStyle(m.group(1).toInt) : HorizonsDesignation.Asteroid, m.group(2)))
+            } \/> "Could not match '4 Vesta' header pattern."
+
+          // First one that works!
+          case0 orElse
+          case1 orElse
+          case2 orElse "Could not parse the header line.".left
+
+
+        case Search.MajorBody(_) =>
+
+          // Common case is that we have many results, or none.
+          lazy val case0 = 
+            parseMany[Row[HorizonsDesignation.MajorBody]](header, tail, """Multiple major-bodies match string""".r) { os =>
+              (os.lift(0) |@| os.lift(1)).tupled.map {
+                case ((ors, ore), (ons, one)) => { row =>
+                  val rec  = row.substring(ors, ore).trim.toInt
+                  val name = row.substring(ons, one).trim
+                  Row(HorizonsDesignation.MajorBody(rec.toInt), name)
+                }
+              }
+            } .map(_.filterNot(_.a.num < 0)) // filter out spacecraft
+
+          // Single result with form:  Revised: Aug 11, 2015       Charon / (Pluto)     901
+          lazy val case1 =
+            """  +(.*?)  +(\d+) $""".r.findFirstMatchIn(header).map { m =>
+              List(Row(HorizonsDesignation.MajorBody(m.group(2).toInt), m.group(1)))
+            } \/> "Could not match 'Charon / (Pluto)     901' header pattern."
+
+          // First one that works, otherwise Nil because it falls through to small-body search
+          case0 orElse 
+          case1 orElse Nil.right
+      
+      }
+  }
+
+  // Construct a program thet performs a HORIZONS request and yields the response as lines of text.
+  private def horizonsRequestLines(params: Map[String, String]): HS2[List[String]] =
+    horizonsRequest(params) { method =>
+      IO {
+        val s = Source.fromInputStream(method.getResponseBodyAsStream, method.getRequestCharSet)
+        try     s.getLines.toList
+        finally s.close()
+      }
+    }
+
+  implicit def GetMethodResource: Resource[GetMethod] =
+    new Resource[GetMethod] {
+      def close(m: GetMethod): IO[Unit] =
+        IO(m.releaseConnection)
+    }
+
+  // Construct a program thet performs a HORIZONS request and transforms the respons with the
+  // provided function.
+  private def horizonsRequest[A](params: Map[String, String])(f: GetMethod => IO[A]): HS2[A] =
+    EitherT {
+      IO(new GetMethod(CgiHorizonsConstants.HORIZONS_URL)).using { (method: GetMethod) =>
+        IO {
+          HorizonsService2.synchronized {
```

Comment:
Yep, I will add a comment. Strictly speaking I'm not quite sure this is adequate because the `GetMethod` remains open and results could in principle still be streaming back after the monitor is released. However I have not been able to make this happen so I'm going to keep an eye on it. I guess an explicit mutex would be safer.